### PR TITLE
Expose the new leader schedule functionality from the bank.

### PR DESCRIPTION
**Pull request by garious**
#### Problem

We have two leader scheduling algorithms in place. There can be only one.

#### Summary of Changes

Expose the new leader schedule functionality from the bank. The most noteworthy addition is the new `Bank::slot_leader()` method. It returns the leader id of the current slot.  It does it by traversing back to a parent bank to get a snapshot of the stakers and uses that data to generate a schedule for the current epoch. After it has generated the schedule, it simply indexes it by current slot.

Also, we'll eventually want to memoize the schedules, but don't need to do that right away. The schedule is generated from a frozen bank and is generated deterministically. The only motivation to memoize would be to improve the performance of code that awkwardly calls `slot_leader()` many times per slot.

Update! To collect transaction fees, we added a Bank::leader field, set from Bank::new_from_parent(). This PR removes that parameter and instead calculates the leader internally using the new schedule.

Towards #2657

### CODE REVIEW

**Comment by garious** [comment]
Hey @carllin, I stole your vote_states() implementation!

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -675,14 +686,82 @@ impl Bank {
...
     /// Return the checkpointed bank that should be used to generate a leader schedule.
     /// Return None if a sufficiently old bank checkpoint doesn't exist.
-    pub fn leader_schedule_bank(&self) -> Option<Arc<Bank>> {
-        let epoch_slot_height = self.slot_height() - self.slot_index();
+    fn leader_schedule_bank(&self, epoch_height: u64) -> Option<Arc<Bank>> {
+        let epoch_slot_height = epoch_height * self.slots_per_epoch();
         let expected = epoch_slot_height.saturating_sub(self.leader_schedule_slot_offset);
         self.parents()
             .into_iter()
             .find(|bank| bank.slot_height() <= expected)
```
Comment (carllin) [inline runtime/src/bank.rs]: The slot_leader() function may be be called after the bank at `epoch_slot_height.saturating_sub(self.leader_schedule_slot_offset);` is merged into the root, in which case this bank for generating for leader schedule will no longer be found. Will this be addressed in a different PR where we do memoization?

Comment (garious) [reply runtime/src/bank.rs]: Yes, separate PR.  I see 3 options: memoize at each epoch; memoize only in the root; reject some attempts to merge as "too soon".  The first option seems most likely.

**Comment by garious** [comment]
Last call for reviews!

File: runtime/src/bank.rs (hunk 1 above)
Comment (garious) [reply runtime/src/bank.rs]: @carllin, I updated the code so that we'd only need to memoize `staked_nodes()`, not the leader schedule. Figured that might have utility elsewhere (@sagar-solana, gossip?), and it let me reduce the LeaderSchedule to almost no code.

**Comment by garious** [comment]
@sagar-solana, I pushed that code back into LeaderScheduler. I'm now using your HashMap. I also made a bunch of recent changes to this PR to make this more accessible to your code. Check it out!

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -673,14 +683,90 @@ impl Bank {
...
 
-    /// Return the checkpointed bank that should be used to generate a leader schedule.
-    /// Return None if a sufficiently old bank checkpoint doesn't exist.
-    pub fn leader_schedule_bank(&self) -> Option<Arc<Bank>> {
-        let epoch_slot_height = self.slot_height() - self.slot_index();
-        let expected = epoch_slot_height.saturating_sub(self.leader_schedule_slot_offset);
-        self.parents()
-            .into_iter()
+    /// Return the checkpointed stakes that should be used to generate a leader schedule.
+    fn staked_nodes_at_epoch(&self, epoch_height: u64) -> HashMap<Pubkey, u64> {
```
Comment (garious) [inline runtime/src/bank.rs]: @sagar-solana, this is for you!

**Comment by garious** [comment]
Post-merge reviews graciously accepted
